static/js/settings_emoji.js

```diff
@@ -159,11 +172,129 @@ export function build_emoji_upload_widget() {
+        $preview_image.show();
+    });
+
+    $preview_text.show();
+    $clear_button.on("click", (e) => {
+        e.stopPropagation();
+        e.preventDefault();
+
+        $("#add-custom-emoji-modal .dialog_submit_button").prop("disabled", true);
+
+        $preview_image.hide();
+        $placeholder_icon.show();
+        $preview_text.show();
+        $preview_image.attr("src", "/static/images/default-avatar.png");
+    });
+}
+
+function show_modal() {
+    const html_body = render_add_emoji();
+
+    function add_custom_emoji(e) {
+        e.preventDefault();
+        e.stopPropagation();
+
+        dialog_widget.show_dialog_spinner();
+
+        const $emoji_status = $("#dialog_error");
+        const emoji = {};
+
+        function submit_custom_emoji_request(formData) {
+            channel.post({
+                url: "/json/realm/emoji/" + encodeURIComponent(emoji.name),
+                data: formData,
+                cache: false,
+                processData: false,
+                contentType: false,
+                success() {
+                    dialog_widget.close_modal();
+                },
+                error(xhr) {
+                    $("#dialog_error").hide();
+                    dialog_widget.hide_dialog_spinner();
+                    const errors = JSON.parse(xhr.responseText).msg;
+                    xhr.responseText = JSON.stringify({msg: errors});
+                    ui_report.error($t_html({defaultMessage: "Failed"}), xhr, $emoji_status);
+                },
+            });
+        }
+
+        for (const obj of $("#add-custom-emoji-form").serializeArray()) {
+            emoji[obj.name] = obj.value;
+        }
+
+        if (emoji.name.trim() === "") {
+            ui_report.client_error(
+                $t_html({defaultMessage: "Failed: Emoji name is required."}),
+                $emoji_status,
+            );
+            dialog_widget.hide_dialog_spinner();
+            return;
+        }
+
+        if (is_custom_emoji(emoji.name)) {
+            ui_report.client_error(
+                $t_html({
+                    defaultMessage: "Failed: A custom emoji with this name already exists.",
+                }),
+                $emoji_status,
+            );
+            dialog_widget.hide_dialog_spinner();
+            return;
+        }
+
+        const formData = new FormData();
+        for (const [i, file] of Array.prototype.entries.call($("#emoji_file_input")[0].files)) {
+            formData.append("file-" + i, file);
+        }
+
+        if (is_default_emoji(emoji.name)) {
+            if (!page_params.is_admin) {
+                ui_report.client_error(
+                    $t_html({
+                        defaultMessage:
+                            "Failed: There is a default emoji with this name. Only administrators can override default emoji.",
+                    }),
+                    $emoji_status,
+                );
+                return;
```

Comment:
This block is missing a `dialog_widget.hide_dialog_spinner`; the other code paths all look good.